// node-comments/index.js
/**
 * Created by thimpat on 03/11/2015.
 */

/*
// That was for the middleware
module.exports = function() {

    return function(req, res, next)
    {
        console.log(req.body);
        console.log(req.query);
        next();
    }
};
*/

(function()
{
    'use strict';

    var express,
        router,
        config,
        jsonfile,
        path,
        fileExists;

    express = require('express');
    router = express.Router();
    jsonfile = require('jsonfile');
    fileExists = require('file-exists');
    path = require('path');

    jsonfile.spaces = 4;

    function sendError(res)
    {
        res.status(400);
        res.json({data: 'failed'});
    }

    function convertToId(idpage)
    {
        var arr = []
        arr.push("./tmp/" + getUrlParameter(idpage,"id"));
        arr.push("/" + getUrlParameter(idpage,"testid") + "/comments-data.js");

        var arr1 = []
        arr1.push("./" + getUrlParameter(idpage,"testid") + "/comments-data.js");

        var abfilepath = path.resolve(arr.join(""));
        var  main_str = abfilepath;
        var  sub_str  = 'tmp\\';

        var subStr = sub_str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        var ct = (main_str.match(new RegExp(subStr, 'gi')) || []).length;
        if(ct > 1 )
        {
            return arr1.join("");
        }
        else
        {
            return arr.join("");
        }
    }

    function getUrlParameter(url, name) {
    name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]');
    var regex = new RegExp('[\\?&]' + name + '=([^&#]*)');
    var results = regex.exec(url);
    return results === null ? '' : decodeURIComponent(results[1].replace(/\+/g, ' '));
    }

    function getFilepathFromID(idpage)
    {
        var filepath, filename;
        idpage = convertToId(idpage);

        filepath = idpage;
        //filepath = path.join(config.dbpath, filename + '.json');
        return filepath;
    }

    function getData(idpage)
    {
        var filepath,
           dataComments;
        //filepath = config.dbpath;
        filepath = getFilepathFromID(idpage);

        var abfilepath = path.resolve(filepath);
        dataComments = jsonfile.readFileSync(abfilepath) || [];
        return dataComments;
    }


    function saveData(idpage, obj, successCb, errorCb)
    {
        var filepath;
        filepath = getFilepathFromID(idpage);
        jsonfile.writeFile(filepath, obj, function (err)
        {
            if (err)
            {
                errorCb && errorCb();
                return false;
            }
            successCb && successCb();
        });
    }

    router.get('/all', function(req, res, next){
        var all,
            idpage;
        idpage = req.headers.referer;
        all = getData(idpage);
        if (config.mongodb)
        {
            // Save to MongoDB database
        }
        else
        {
            res.json(all);
        }
    });

    router.post('/add', function (req, res, next)
    {
        var idPage,
            all;
        if (config.mongodb)
        {
            // Save to MongoDB database
        }
        else
        {
            // Save to file
            idPage = req.headers.referer;

            all = getData(idPage);
            all.push(req.body);
            saveData(
                idPage,
                all,
                function()
                {
                    res.json({data: 'success'});
                },
                function()
                {
                    sendError(res);
                }
            );
        }
    });

    router.post('/update', function(req, res, next){
        var all,
            n, i, k, id, idpage;
        id = req.body.id;

        idpage = req.headers.referer;

        if (config.mongodb)
        {
            // Save to MongoDB database
        }
        else
        {
            all = getData(idpage);
            n = all.length;
            for (i = 0; i < n; ++i)
            {
                k = all[i];
                if (id === k.id)
                {
                    all[i] = req.body;
                    saveData(
                        idpage,
                        all,
                        function()
                        {
                            res.json({data: 'success'});
                        },
                        function()
                        {
                            sendError(res);
                        });
                    break;
                }
            }
        }
    });

    router.post('/delete', function(req, res, next)
    {
        var all,
            n, i, k, id, idpage;
        id = req.body.id;
        idpage = req.headers.referer;

        if (config.mongodb)
        {
            // Save to MongoDB database
        }
        else
        {
            all = getData(idpage);
            n = all.length;
            for (i = n - 1; i >= 0; --i)
            {
                k = all[i];
                if (id === k.id)
                {
                    all.splice(i, 1);
                    continue;
                }

                if (id === k.parent)
                {
                    all.splice(i, 1);
                }
            }

            saveData(idpage, all,
                function()
                {
                    res.json({data: 'success'});
                },
                function()
                {
                    sendError(res);
                });
        }
    });

    module.exports.init = function (params)
    {
        config = params;
        config.dbpath = config.dbpath || './';
    };

    module.exports.routes = router;

}());
